src/site/content/en/blog/racer/index.md

```diff
@@ -0,0 +1,289 @@
+
+### Calculating Latency
+
+The starting point for syncing devices is knowing how long it takes for messages to be received from the Compute Engine relay. The tricky part is that the clocks on each device will never be completely in sync. To get around this, we needed to find the difference in time between the device and the server.
+
+To find the time offset between the device and the main server, we send a message with the current device timestamp. The server will then reply with the original timestamp along with the timestamp of the server. We use the response to calculate the actual difference in time.
+
+```js
+var currentTime = Date.now();
+var latency = Math.round((currentTime - e.time) * .5);
+var serverTime = e.serverTime;
+currentTime -= latency;
+var difference = currentTime - serverTime;
+```
+
+Doing this once is not enough, as the round trip to the server is not always symmetrical, meaning it may take longer for the response to get to the server than it does for the server to return it. To get around this, we poll the server multiple times, taking the median result. This gets us within 10ms of the actual difference between device and server.
+
+### Acceleration/Deceleration
+
+When Player 1 presses or releases the screen, the acceleration event is sent to the server. Once received, the server adds its current timestamp and then passes that data along to every other player.
+
+When an “accelerate on” or “accelerate off” event is received by a device, we are able to use the server offset (calculated above) to find out how long it took for that message to be received. This is useful, because Player 1 may receive the message in 20ms, but Player 2 might take 50ms to receive it. This would result in the car being in two different places because device 1 would start the acceleration sooner.
+
+We can take the time it took to receive the event and convert that into frames. At 60fps, each frame is 16.67ms -- so we can add more velocity (acceleration) or friction (deceleration) on of the car to account for the frames it missed.
+
+```js
+var frames = time / 16.67;
+var onScreen = this.isOnScreen() && time < 75;
+for (var i = 0; i < frames; i++) {
+    if (onScreen) {
+        _velocity.length += _throttle * Math.round(frames*.215);
+    } else {
+        _this.render();
+    }
+}
+```
+
+In the example above, if Player 1 has the car on its screen and the time it took to receive the message is less than 75ms, it will adjust the car’s velocity, speeding it up to make up the difference. If the device is not on screen or the message took too long, it will run the render function and actually make the car jump to where it needs to be.
```

Comment:
```suggestion
In the example above, if Player 1 has the car on its screen and the time it took to receive the message is less than 75ms, it will adjust the car's velocity, speeding it up to make up the difference. If the device is not on screen or the message took too long, it will run the render function and actually make the car jump to where it needs to be.
```